DIPs/1NNN-MHH.md

```diff
@@ -0,0 +1,127 @@
+
+## Abstract
+
+This DIP unifies the syntax of function literals and function definitions.
+The following syntax is proposed: 
+```D
+int add(int x, int y) pure => x + y;
+```
+
+Thanks to a failure of process, this feature is already present inside the implementation of the 
+D programming language. This DIP will disambiguate the status of this feature, i.e. after thorough 
+consideration a decision can be made and acted upon rather than leaving features in limbo.
+
+## Contents
+* [Rationale](#rationale)
+* [PriorWork](#prior-work)
+* [Description](#description)
+* [Reference](#reference)
+* [Copyright & License](#copyright--license)
+* [Reviews](#reviews)
+
+## Rationale
+There already exists [syntactic sugar](https://en.wikipedia.org/wiki/Syntactic_sugar) for function literals in the D programming language, for example
+```D
+const succ = (int x) => x + 1;
+```
+is equivalent to
+```D
+const succ = function(int x) { return x + 1; };
+```
+No such syntactic sugar exists for functions and method definitions - therefore, a trivial change to the language unifies the syntax of 
+those constructs and saves a little typing for the programmer.
+
+## Prior Work
+As mentioned previously, this DIP has the potential to result in a post-facto blessing of 
+a feature already present (available through a preview flag) in the compiler, this implementation
+was written by Adam Ruppe - in recognition of this he is listed as a coauthor.
+
+This feature is present within the C# programming language, where instances of its use are referred to as 
+[Expression-bodied members](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/statements-expressions-operators/expression-bodied-members).
+This feature has been available for use since either C# version 6 or 7 depending on context - for reference the most recent stable release is version 9.0.
+
+The C# documentation linked above gives an example of idiomatic use of this feature (reproduced below). In considering this piece of code, we 
+should be careful to remember that C# (to some degree of approximation at least) encourages access to state through methods (Getters and Setters, properties, etc.),
+so the use of this feature may be more profitable in C# than in D where the aforementioned idiom is not as popular - that is to say that in the authors experience 
+D programs with many single line functions bodies are significantly less prevalent than in C#.
+```c#
+public class Location
+{
+   private string locationName;
+
+   public Location(string name)
+   {
+      locationName = name;
+   }
+
+   public string Name => locationName;
+}
+```
+## Description
+### Syntax 
+The grammar changes are as follows:
+```diff
+FunctionBody:
+     SpecifiedFunctionBody
+     MissingFunctionBody
++    ShortenedFunctionBody
+...
+
++ShortenedFunctionBody:
++    => AssignExpression ;
+```
+### Semantics 
+This proposed semantics are a simple example of what is often referred to as lowering:
```

Comment:
Not a language expert but I don't think this is a case of lowering. This describes additional grammar, and not how to transform ASTs.

Reply:
I think 'lowering' works fine here. Whether it's happening during semantic analysis or parsing, it's still rewriting a specific construct into a more general construct so you don't have to think about the more specific construct anymore.

Reply:
Makes no difference either way - "lowering" is a handwavy term which to my knowledge may have first been properly documented via the D community (https://stackoverflow.com/questions/20252876/wanted-good-definition-of-the-term-lowering-in-the-context-of-compilers). All that matters is that the semantics of this addition are defined exactly in terms of things already defined within the language rather than ab initio - this doesn't dictate the *implementation* (even though it should) because programming language specifications are just that - i.e. a dictation of behaviour, or more specifically enough just enough behaviour for the specification to enable reasoning about programs. See the famous "as-if" rule for example.